### PR TITLE
[SPARK-7727] Avoid inner classes in RuleExecutor

### PR DESCRIPTION
**Pull request by Jeffrharr**
Allow composibility of Batch in RuleExecutor be making classes outer.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/package.scala [hunk 1]
```diff
@@ -19,6 +19,7 @@ package org.apache.spark.sql.catalyst.expressions
 
 import org.apache.spark.annotation.DeveloperApi
 import org.apache.spark.sql.catalyst.rules
+import org.apache.spark.sql.catalyst.rules._
```
Comment (smola) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/package.scala]: This shouldn't be needed.

**Comment by chenghao-intel** [comment]
Any background why we should expose the `Batch`? Per my understanding, we'd better keep the minimizing the class visibility as much as possible in design.

**Comment by His-name-is-Joof** [comment]
In my opinion, it seems like a poor fix to the issue of composing Batches
in different instances and there is likely a much better way to refactor
the class to get the desired behavior (although I don't personally know how
yet). It was labeled as starter and making the classes outer was the
proposed fix.

**Comment by chenghao-intel** [comment]
Oh, sorry, I didn't see the description of the jira. I will jump there to give more comments.

**Comment by evacchi** [comment]
@chenghao-intel although Batch may not be supposed to be public, inner classes are still bad when it comes to sharing code (because they are not static inner). I'd advise to make them at least inner to a companion object; see PR #6349

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 2]
```diff
@@ -61,7 +62,7 @@ class Analyzer(
    */
   val extendedResolutionRules: Seq[Rule[LogicalPlan]] = Nil
 
-  lazy val batches: Seq[Batch] = Seq(
+  lazy val batches = Seq(
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: Can you add the explicit type back?

**Comment by rxin** [comment]
LGTM other than that explicit type.

**Comment by rxin** [comment]
Jenkins, test this please.

**Comment by marmbrus** [comment]
ping, mind bringing this up to date?

**Comment by rxin** [comment]
Jenkins, ok to test.

**Comment by marmbrus** [comment]
ok to test

**Comment by marmbrus** [comment]
@His-name-is-Joof, thanks for working on this.  I propose we close this issue until you have time to fix the build failures and bring the code up to date.

**Pull request closed without merging**